File: CubeController/pages/ListMenuPage.py
from src import DisplayManager
from pages import Page

class ListMenuPage(Page):
    def __init__(self, page_manager: "PageManager", display_manager: DisplayManager):
        super().__init__(page_manager, display_manager)
        self.items = [
            "Manual setting",
            "Face detection",
            "Solve Cube"
        ]
        self.selected_index = 0
    
    def show(self):
        self.display_manager.clear()
        for i, item in enumerate(self.items):
            y = 5 + i * 20
            self.display_manager.write(f"{'>' if i == self.selected_index else ' '}{item}", y)
            
        print("Showing List")
        self.display_manager.show()
    
    def increment(self, delta: int):
        self.selected_index = (self.selected_index + delta) % len(self.items)
        self.show()
    
    def press(self):
        print(f"Pressed {self.items[self.selected_index]}")

File: CubeController/src/pages/ListMenuPage.cpp
#include "pages/ListMenuPage.h"
#include "DisplayManager.h"

void ListMenuPage::show() {
    DisplayScope scope;
    int y = 20;
    int lineHeight = DisplayManager::getLineHeight();
    for(int i = 0; i < items.size(); i++) {
        String name = items[i].name;
        DisplayManager::Style style = i == selectedItem ? DisplayManager::INVERTED : DisplayManager::NORMAL;
        DisplayManager::print(name.c_str(), 0, y, style);
        y += lineHeight + 5;
    }
}

void ListMenuPage::increment(int delta) {
    selectedItem = (selectedItem + delta + items.size()) % items.size();
    show();
}

std::optional<PagesEnum> ListMenuPage::press() {
    if(selectedItem < 0 || selectedItem >= items.size()) {
        return std::nullopt;
    }
    PagesEnum page = items[selectedItem].page;
    return page;
}